src/transformers/generation_tf_utils.py

```diff
@@ -1653,12 +1655,25 @@ def _expand_to_num_beams(tensor: tf.Tensor, num_beams: int) -> tf.Tensor:
     def _prepare_attention_mask_for_generation(
         self,
         inputs: tf.Tensor,
-        pad_token_id: int,
+        pad_token_id: Optional[int],
+        eos_token_id: Optional[int],
     ) -> tf.Tensor:
         is_input_ids = len(inputs.shape) == 2 and inputs.dtype in (tf.int32, tf.int64)
         is_pad_token_in_inputs = (pad_token_id is not None) and tf.math.reduce_any(inputs == pad_token_id)
+        # some models, like GPT-2, don't set a pad token, and it gets defaulted to eos_token_id (we don't want to pick
+        # up on those)
+        is_pad_token_not_equal_to_eos_token_id = (eos_token_id is None) or (
+            (eos_token_id is not None) and (pad_token_id != eos_token_id)
+        )
+        # however, they can also have left-padding (we definitely want to pick up on those)
+        has_trailing_pad_tokens = tf.math.reduce_any(inputs[:, 0] == pad_token_id)
+
         # Check if input is input_ids and padded -> only then is attention_mask defined
-        if is_input_ids and is_pad_token_in_inputs:
+        if (
+            is_input_ids
+            and is_pad_token_in_inputs
+            and (has_trailing_pad_tokens or is_pad_token_not_equal_to_eos_token_id)
+        ):
```

Comment:
@patrickvonplaten I think I found a bug in PT generation.

The `is_pad_token_not_equal_to_eos_token_id` condition exists in PT, it is copied here. It is used to infer the attention mask correctly when `eos_token_id==pad_token_id`, so that the eos doesn't get masked. However, GPT-2 also uses left-padding for generation (and we set `pad_token_id=eos_token_id`, because the former is missing). Without the new `has_trailing_pad_tokens`, the infered attention mask is incorrect when left-padding is present.

An example running from this branch, where TF passes the check and PT doesn't:
```
from transformers import GPT2Tokenizer, TFGPT2LMHeadModel, GPT2LMHeadModel

tokenizer = GPT2Tokenizer.from_pretrained("distilgpt2")
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "left"


tf_model = TFGPT2LMHeadModel.from_pretrained("distilgpt2")

tf_inputs = tokenizer(["Hello"], padding=True, return_tensors="tf")
tf_generation_output = tf_model.generate(tf_inputs.input_ids)
tf_outputs = tokenizer.batch_decode(tf_generation_output, skip_special_tokens=True)

tf_inputs_left_padded = tokenizer(["Hello", "Long input to trigger left padding"], padding=True, return_tensors="tf")
tf_generation_output_left_padded = tf_model.generate(tf_inputs_left_padded.input_ids)
tf_outputs_left_padded = tokenizer.batch_decode(tf_generation_output_left_padded, skip_special_tokens=True)
assert(tf_outputs[0][:60] == tf_outputs_left_padded[0][:60])


pt_model = GPT2LMHeadModel.from_pretrained("distilgpt2")

pt_inputs = tokenizer(["Hello"], padding=True, return_tensors="pt")
pt_generation_output = pt_model.generate(pt_inputs.input_ids)
pt_outputs = tokenizer.batch_decode(pt_generation_output, skip_special_tokens=True)

pt_inputs_left_padded = tokenizer(["Hello", "Long input to trigger left padding"], padding=True, return_tensors="pt")
pt_generation_output_left_padded = pt_model.generate(pt_inputs_left_padded.input_ids)
pt_outputs_left_padded = tokenizer.batch_decode(pt_generation_output_left_padded, skip_special_tokens=True)
assert(pt_outputs[0][:60] == pt_outputs_left_padded[0][:60])
```

Shall I also propagate the change to PT?

Reply:
Let's discuss in the call quickly :-)

Reply:
As discussed let's rather throw a warning here :-)